### PR TITLE
Tma 633 added validation for sync_mode

### PR DESCRIPTION
**Pull request by Hahihula**

### CODE REVIEW

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 1]
```diff
@@ -143,4 +143,34 @@
       end
     end
   end
+
+  context 'when using unsuported sync_mode' do
+    let(:params) do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Similar code found in synchronize_user_filters_spec.rb:79, synchronize_user_filters_spec.rb:148 (mass = 52)

File: spec/unit/actions/synchronize_users_spec.rb [hunk 2]
```diff
@@ -130,5 +130,28 @@
         let(:message_for_project) { :import_users }
       end
     end
+
+    context 'when using unsuported mode' do 
+      let(:params) do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Similar code found in synchronize_users_spec.rb:47, synchronize_users_spec.rb:76, synchronize_users_spec.rb:135 (mass = 60)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 3]
```diff
@@ -63,7 +74,10 @@ def call(params)
           symbolized_config[:labels] = symbolized_config[:labels].map { |l| GoodData::Helpers.symbolize_keys(l) }
           headers_in_options = params.csv_headers == 'false' || true
 
-          mode = params.sync_mode || 'sync_project'
+          mode = params.sync_mode
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Refers to 'params' more than self (maybe move it to another class?) (FeatureEnvy)

File: lib/gooddata/lcm/actions/synchronize_user_filters.rb [hunk 4]
```diff
@@ -63,7 +74,10 @@ def call(params)
...
+          unless mode.nil? || MODES.include?(mode)
```
Comment (yenkins) [inline lib/gooddata/lcm/actions/synchronize_user_filters.rb]: Performs a nil-check (NilCheck)

File: spec/unit/actions/synchronize_users_spec.rb [hunk 5]
```diff
@@ -130,5 +130,28 @@
         let(:message_for_project) { :import_users }
       end
     end
+
+    context 'when using unsuported mode' do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Style/TrailingWhitespace: Trailing whitespace detected.

File: spec/unit/actions/synchronize_users_spec.rb [hunk 6]
```diff
@@ -130,5 +130,28 @@
...
+          gdc_logger: logger,
+          sync_mode: 'unsuported_sync_mode'
+        }
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      before do
+        allow(File).to receive(:open).and_return("client_id\n123456789")
+      end
+
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Style/TrailingWhitespace: Trailing whitespace detected.

File: spec/unit/actions/synchronize_users_spec.rb [hunk 7]
```diff
@@ -130,5 +130,28 @@
...
+      before do
+        allow(File).to receive(:open).and_return("client_id\n123456789")
+      end
+        
+      it 'fails' do
+        expect { subject.class.call(params) }.to raise_error
+      end
+
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Style/EmptyLinesAroundBlockBody: Extra empty line detected at block body end.

File: spec/unit/actions/synchronize_users_spec.rb [hunk 8]
```diff
@@ -130,5 +130,28 @@
...
+    end
+
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Style/EmptyLinesAroundBlockBody: Extra empty line detected at block body end.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/684/ : SUCCESS in 14s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/511/ : FAILURE in 41s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/795/ : SUCCESS in 2m 07s

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 9]
```diff
@@ -162,11 +161,39 @@
...
     end
-
     it 'fails if the MUF set is empty' do
       expect(File).to receive(:open)
       expect(CSV).to receive(:foreach)
       expect { subject.class.call(params) }.to raise_error(/The filter set can not be empty/)
     end
   end
+  context 'when using unsuported sync_mode' do
+    let(:params) do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Similar code found in synchronize_user_filters_spec.rb:94, synchronize_user_filters_spec.rb:171 (mass = 52)

File: spec/unit/actions/synchronize_user_filters_spec.rb [hunk 10]
```diff
@@ -162,11 +161,39 @@
...
+        GDC_GD_CLIENT: client,
+        input_source: 'foo',
+        domain: 'bar',
+        filters_config: { labels: [] },
+        multiple_projects_column: 'id_column',
+        sync_mode: 'unsuported_sync_mode',
+        gdc_logger: logger
+      }
+      GoodData::LCM2.convert_to_smart_hash(params)
+    end
```
Comment (yenkins) [inline spec/unit/actions/synchronize_user_filters_spec.rb]: Style/TrailingWhitespace: Trailing whitespace detected.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/685/ : SUCCESS in 11s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/514/ : FAILURE in 35s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/796/ : SUCCESS in 34s

File: spec/unit/actions/synchronize_users_spec.rb [hunk 11]
```diff
@@ -130,5 +130,23 @@
         let(:message_for_project) { :import_users }
       end
     end
+    context 'when using unsuported mode' do
+      let(:params) do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Similar code found in synchronize_users_spec.rb:47, synchronize_users_spec.rb:76, synchronize_users_spec.rb:134 (mass = 60)

File: spec/unit/actions/shared_examples_for_user_actions.rb [hunk 12]
```diff
@@ -24,3 +24,20 @@
...
+  before do
+      allow(project).to receive(:metadata).and_return(
+        'GOODOT_CUSTOM_PROJECT_ID' => 'project-123'
+      )
+      allow(project).to receive(:uri).and_return('project-uri')
+      allow(project).to receive(:add_data_permissions)
+      allow(domain).to receive(:clients).and_return([])
+      allow(data_source).to receive(:realize).and_return('filepath')
+      allow(File).to receive(:open).and_return("client_id\n123456789")
+    end
```
Comment (yenkins) [inline spec/unit/actions/shared_examples_for_user_actions.rb]: Lint/BlockAlignment: `end` at 38, 4 is not aligned with `before do` at 29, 2.

File: spec/unit/actions/shared_examples_for_user_actions.rb [hunk 13]
```diff
@@ -24,3 +24,20 @@
...
+
+  it 'fails' do
+    expect { subject.class.call(params) }.to raise_error
+  end
+end
```
Comment (yenkins) [inline spec/unit/actions/shared_examples_for_user_actions.rb]: Style/TrailingBlankLines: Final newline missing.

File: spec/unit/actions/synchronize_users_spec.rb [hunk 14]
```diff
@@ -130,5 +130,23 @@
...
+          gdc_logger: logger,
+          sync_mode: 'unsuported_sync_mode' # sync_one_project_based_on_pid
+        }
+        GoodData::LCM2.convert_to_smart_hash(params)
+      end
+
+      before do
+        allow(domain).to receive(:clients).and_return(organization)
+      end
+
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Style/TrailingWhitespace: Trailing whitespace detected.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/686/ : SUCCESS in 12s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/516/ : FAILURE in 43s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/797/ : SUCCESS in 36s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/687/ : SUCCESS in 10s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/517/ : FAILURE in 37s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/798/ : SUCCESS in 39s

File: spec/unit/actions/synchronize_users_spec.rb [hunk 15]
```diff
@@ -130,5 +130,22 @@
         let(:message_for_project) { :import_users }
       end
     end
+    context 'when using mistyped mode' do
+      let(:params) do
```
Comment (yenkins) [inline spec/unit/actions/synchronize_users_spec.rb]: Similar code found in synchronize_users_spec.rb:76, synchronize_users_spec.rb:134 (mass = 40)

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/688/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/518/ : FAILURE in 40s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/799/ : SUCCESS in 35s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/689/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/519/ : FAILURE in 36s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/800/ : SUCCESS in 38s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/690/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/520/ : SUCCESS in 30s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/801/ : SUCCESS in 33s

File: spec/unit/actions/shared_examples_for_user_actions.rb [hunk 16]
```diff
@@ -24,3 +24,20 @@
...
+    )
+    allow(project).to receive(:uri).and_return('project-uri')
+    allow(project).to receive(:add_data_permissions)
+    allow(domain).to receive(:clients).and_return([])
+    allow(data_source).to receive(:realize).and_return('filepath')
+    allow(File).to receive(:open).and_return("client_id\n123456789")
+  end
+
+  it 'fails' do
+    expect { subject.class.call(params) }.to raise_error
```
Comment (kubamahnert) [inline spec/unit/actions/shared_examples_for_user_actions.rb]: Maybe we could check the exception message with a regex to make sure it is the one we expected?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/691/ : SUCCESS in 9s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/521/ : SUCCESS in 31s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/802/ : SUCCESS in 32s

**Comment by gdgate** [comment]
Merge Failed.

This change or one of its cross-repo dependencies was unable to be automatically merged with the current state of its repository. Please rebase the change and upload a new patchset.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://ci.intgdc.com/job/gooddata-ruby-opensource-repository-check/692/ : SUCCESS in 8s
- https://ci.intgdc.com/job/gooddata-ruby-pronto-code-review-zuul/522/ : SUCCESS in 31s
- https://ci.intgdc.com/job/gooddata-ruby-unit-tests-zuul/803/ : SUCCESS in 1m 56s

**Pull request closed without merging**